clients/pubsub-common/src/main/java/org/eclipse/hono/client/pubsub/subscriber/PubSubSubscriber.java

```diff
@@ -0,0 +1,117 @@
+     * based on a created subscription, which follows the format: projects/{project}/subscriptions/{subscription}
+     *
+     * @param vertx The Vert.x instance that this subscriber runs on.
+     * @param projectId The identifier of the Google Cloud Project to connect to.
+     * @param subscriptionId The name of the subscription to create the subscriber for.
+     * @param receiver The message receiver used to process the received message.
+     * @param credentialsProvider The provider for credentials to use for authenticating to the Pub/Sub service.
+     * @throws NullPointerException If any of these parameters is {@code null}.
+     */
+    public PubSubSubscriber(
+            final Vertx vertx,
+            final String projectId,
+            final String subscriptionId,
+            final MessageReceiver receiver,
+            final FixedCredentialsProvider credentialsProvider) {
+        this.vertx = Objects.requireNonNull(vertx);
+        Objects.requireNonNull(projectId);
+        Objects.requireNonNull(subscriptionId);
+        Objects.requireNonNull(receiver);
+        Objects.requireNonNull(credentialsProvider);
+
+        final ProjectSubscriptionName subscriptionName = ProjectSubscriptionName.of(projectId, subscriptionId);
+        this.subscriber = Subscriber
+                .newBuilder(subscriptionName, receiver)
+                .setCredentialsProvider(credentialsProvider)
+                .build();
+    }
+
+    /**
+     * Subscribes messages from Pub/Sub.
+     *
+     * @return A future indicating the outcome of the operation.
+     * @throws ServerErrorException If subscribing was not successful.
+     */
+    public Future<Void> subscribe() {
+        try {
+            subscriber.addListener(
+                    new ApiService.Listener() {
+
+                        @Override
+                        public void failed(final ApiService.State from, final Throwable failure) {
+                            log.error("Error subscribing message from Pub/Sub", failure);
+                            throw new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE,
+                                    "Error subscribing message from Pub/Sub", failure);
+                        }
+                    },
+                    MoreExecutors.directExecutor());
+            subscriber.startAsync().awaitRunning();
+            return Future.succeededFuture();
+        } catch (IllegalStateException e) {
+            log.error("Service reached illegal state", e);
+            return Future.failedFuture(e);
+        }
+    }
+
+    /**
+     * Shuts the subscriber down and frees resources.
+     */
+    @Override
+    public void close() {
+        vertx.executeBlocking(result -> {
+            if (subscriber == null) {
+                result.complete();
+            } else {
+                subscriber.stopAsync();
```

Comment:
it feels strange that we should need to wrap this into *executeBlocking*. Does this method actually wait for the stopping to have finished?

Reply:
No, removed the executeBlocking